### DISCUSSION TITLE
How to deal with concurrency

### DISCUSSION THREAD

**Discussion by nsoubelet** [dCache/nfs4j#126] [Q&A] [answered]
Hello there,
Not sure whether you read this Q&A section, however I'll give it a try.
Basically I implemented a nfs4 to azure files (smb) using this incredible library. However, I am facing a really (basic) issue and I would need to know whether solving it in my implementation FS or there is something in the internals (also what you recommend in this case).

In Azure Files, when opening a file for writing, it is marked as exclusive, meaning that any other operation trying to write into it, will get an error.
However, my problem here is that if I `lock` the write section using any of the standard concurrency mechanisms, I will be penalizing other threads that want to write into different files, which is a perfectly valid operation.

Therefore my question: is there any functionality to enqueue `WriteResult write` operations over the same `Inode` or I should implement that logic in my `VirtualFileSystem` implementation?
In case there is not a standard way, any suggestions on how to approach the problem?

Thanks! :)

**Comment by kofemann** [comment, marked as answer]
Hi,

We are always happy to see that our works are also useful in other projects!

Coming to the questions. NFS4.0 protocol has a concept of [Share reservations](https://datatracker.ietf.org/doc/html/rfc5661#section-9.7), which defines which
concurrent access is allowed. There is `share_access` and `share_deny`. The `share_access`
specifies the access mode which is wanted, i.e. open for read or open for write, while
`deny` defines which concurrent access is allowed, i.e. other readers or other writers.
The POSIX standard doesn't provide a way to specify `share_deny`, thus, typically, the
NFS clients can't specify it. However, nfs4j supports it. The quick-and-dirty workaround
will be to set `_args.opopen.share_deny.value` to `OPEN4_SHARE_DENY_BOTH`
if `share_access` is `OPEN4_SHARE_ACCESS_WRITE`. The NFS server should be updated to
use the new op:

```java
    private static class AzureOperationExecutor extends MDSOperationExecutor {
    
        @Override
        public AbstractNFSv4Operation getOperation(nfs_argop4 op) {
            if (op.argop == nfs_opnum4.OP_OPEN && 
                  op.opopen.share_access.value == OPEN4_SHARE_ACCESS_WRITE) {
               // tweak share deny
               op.opopen.share_deny.value = OPEN4_SHARE_DENY_BOTH;
            }
            return super.getOperation(op);
        }
    }

     NFSServerV41 nfs4 =
        new NFSServerV41.Builder()
            .withOperationExecutor(new AzureOperationExecutor())
            ...
            .build();
```

The code above is of course just an idea and probably should make it right :)

A cleaner solution might be exposing some kind of file system capabilities to define the default behaviour.

I hope this helps.

**Reply by nsoubelet** [reply]
Hello @kofemann, thanks for replying.
That was really helpful. I did not think about adding that logic in the operation executor. 
I will investigate that. :)
